go/vt/vtgate/evalengine/convert.go

```diff
@@ -33,41 +34,95 @@ type (
 
 var ErrConvertExprNotSupported = "expr cannot be converted, not supported"
 
-// translateComparisonOperator takes in a sqlparser.ComparisonExprOperator and
-// returns the corresponding ComparisonOp
-func translateComparisonOperator(op sqlparser.ComparisonExprOperator) ComparisonOp {
+func convertComparisonExpr(op sqlparser.ComparisonExprOperator, left, right sqlparser.Expr, lookup ConverterLookup) (Expr, error) {
+	l, err := convertExpr(left, lookup)
+	if err != nil {
+		return nil, err
+	}
+	r, err := convertExpr(right, lookup)
+	if err != nil {
+		return nil, err
+	}
+	return convertComparisonExpr2(op, l, r)
+}
+
+func convertComparisonExpr2(op sqlparser.ComparisonExprOperator, left, right Expr) (Expr, error) {
+	binaryExpr := BinaryExpr{
+		Left:  left,
+		Right: right,
+	}
+
+	if op == sqlparser.InOp || op == sqlparser.NotInOp {
+		return &InExpr{
+			BinaryExpr: binaryExpr,
+			Negate:     op == sqlparser.NotInOp,
+			Hashed:     nil,
+		}, nil
+	}
+
+	coercedExpr := BinaryCoercedExpr{
+		BinaryExpr: binaryExpr,
+	}
+	if err := coercedExpr.coerce(); err != nil {
+		return nil, err
+	}
+
 	switch op {
 	case sqlparser.EqualOp:
-		return &EqualOp{"=", func(cmp int) bool { return cmp == 0 }}
+		return &ComparisonExpr{coercedExpr, compareEQ{}}, nil
+	case sqlparser.NotEqualOp:
+		return &ComparisonExpr{coercedExpr, compareNE{}}, nil
 	case sqlparser.LessThanOp:
-		return &EqualOp{"<", func(cmp int) bool { return cmp < 0 }}
-	case sqlparser.GreaterThanOp:
-		return &EqualOp{">", func(cmp int) bool { return cmp > 0 }}
+		return &ComparisonExpr{coercedExpr, compareLT{}}, nil
 	case sqlparser.LessEqualOp:
-		return &EqualOp{"<=", func(cmp int) bool { return cmp <= 0 }}
+		return &ComparisonExpr{coercedExpr, compareLE{}}, nil
+	case sqlparser.GreaterThanOp:
+		return &ComparisonExpr{coercedExpr, compareGT{}}, nil
 	case sqlparser.GreaterEqualOp:
-		return &EqualOp{">=", func(cmp int) bool { return cmp >= 0 }}
-	case sqlparser.NotEqualOp:
-		return &EqualOp{"!=", func(cmp int) bool { return cmp != 0 }}
+		return &ComparisonExpr{coercedExpr, compareGE{}}, nil
 	case sqlparser.NullSafeEqualOp:
-		return &NullSafeEqualOp{}
-	case sqlparser.InOp:
-		return &InOp{}
-	case sqlparser.NotInOp:
-		return &InOp{Negate: true}
+		return &ComparisonExpr{coercedExpr, compareNullSafeEQ{}}, nil
 	case sqlparser.LikeOp:
-		return &LikeOp{}
+		return &LikeExpr{BinaryCoercedExpr: coercedExpr}, nil
 	case sqlparser.NotLikeOp:
-		return &LikeOp{Negate: true}
-	case sqlparser.RegexpOp:
-		return &RegexpOp{}
-	case sqlparser.NotRegexpOp:
-		return &RegexpOp{Negate: true}
+		return &LikeExpr{BinaryCoercedExpr: coercedExpr, Negate: true}, nil
 	default:
-		return nil
+		return nil, vterrors.Errorf(vtrpcpb.Code_UNIMPLEMENTED, "%s: %s", ErrConvertExprNotSupported, op.ToString())
 	}
 }
 
+func convertLogicalExpr(opname string, left, right sqlparser.Expr, lookup ConverterLookup) (Expr, error) {
```

Comment:
I'm not quite following why we are sending the op as a string here. wouldn't the original AST struct work?

Reply:
The Logical Expressions in the AST have a different type for each logical expression, so by having this method we can build the resulting Expr node directly. If we pass in the AST node directly, we'd have to switch on it anyway to get the `opname` string, as it is needed when printing the resulting Expr.